### PR TITLE
Use MWC components for ha-form

### PR DESCRIPTION
**Pull request by balloob**
## Proposed change

**To best see impact of this PR, run `gallery/script/develop_gallery`.**

It's time that we start using MWC components instead of Polymer components. This is converting `<ha-form>` to use them. This creates an example for others that want to use these components and it is isolated enough that the different component style can be considered jarring.

However, we should still aim to convert as many places as possible to MWC over Polymer. We should promote this heavily for Hacktoberfest.

image

image

image

To do:

- [x] We only send updates on "onChange" which means the submit button is not "enabled"

Fixed this by always enabling submit button and testing for all required fields when button pressed.

- before:
  image
- after:
  image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by balloob** [comment]
Dark mode still needs some work

image

File: src/components/ha-form/ha-form-boolean.ts [hunk 1]
```diff
@@ -29,23 +30,25 @@ export class HaFormBoolean extends LitElement implements HaFormElement {
...
     });
   }
 
   static get styles(): CSSResultGroup {
     return css`
-      paper-checkbox {
-        display: block;
-        padding: 22px 0;
+      mwc-checkbox {
+        --mdc-theme-secondary: var(--primary-color);
```
Comment (bramkragten) [inline src/components/ha-form/ha-form-boolean.ts:51]: Or use `ha-checkbox`

File: src/components/ha-form/ha-form-integer.ts [hunk 2]
```diff
@@ -31,6 +31,28 @@ export class HaFormInteger extends LitElement implements HaFormElement {
...
+              .max=${this.schema.valueMax}
+              .disabled=${this.data === undefined && this.schema.optional}
+              @change=${this._valueChanged}
+            ></mwc-slider>
+          </div>
+        </div>
+      `;
+    }
+
     return "valueMin" in this.schema && "valueMax" in this.schema
```
Comment (bramkragten) [inline src/components/ha-form/ha-form-integer.ts:56]: The `ha-slider` should be removed now?

File: src/components/ha-form/ha-form-multi_select.ts [hunk 3]
```diff
@@ -40,113 +46,84 @@ export class HaFormMultiSelect extends LitElement implements HaFormElement {
...
-                  ></paper-checkbox>
-                  ${this._optionLabel(item)}
-                </paper-icon-item>
-              `;
-            })
-          }
-        </paper-listbox>
-      </paper-menu-button>
+      <ha-button-menu fixed corner="BOTTOM_START" @closed=${stopPropagation}>
+        <mwc-textfield
```
Comment (bramkragten) [inline src/components/ha-form/ha-form-multi_select.ts:52]: Give the textfield cursor pointer, and add a chevron icon

File: src/resources/styles.ts [hunk 4]
```diff
@@ -13,6 +13,21 @@ export const darkStyles = {
   "switch-unchecked-track-color": "#9b9b9b",
   "divider-color": "rgba(225, 225, 225, .12)",
   "mdc-ripple-color": "#AAAAAA",
+
+  // TODO: adjust these colors for black.
```
Comment (bramkragten) [inline src/resources/styles.ts:17]: Use `--rgb-primary-text-color` as base

File: src/components/ha-form/ha-form-multi_select.ts (hunk 3 above)
Comment (balloob) [reply src/components/ha-form/ha-form-multi_select.ts:52]: I tried, it didn't work. I might have to add it to input element.

Comment (balloob) [reply src/components/ha-form/ha-form-multi_select.ts:104]: Works now.

File: src/resources/styles.ts (hunk 4 above)
Comment (balloob) [reply src/resources/styles.ts]: I just inverted all the colors (255-X)

**Comment by balloob** [comment]
image

File: src/components/ha-form/ha-form-float.ts [hunk 5]
```diff
@@ -25,33 +25,49 @@ export class HaFormFloat extends LitElement implements HaFormElement {
...
+
+    let value: number | undefined;
+
+    if (rawValue !== "") {
+      value = parseFloat(rawValue);
+    }
+
+    // Detect anything changed
+    if (this.data === value) {
+      // parseFloat will drop invalid text at the end, in that case update textfield
```
[the remark this answers is not in the document]
Comment (bramkragten) [reply src/components/ha-form/ha-form-float.ts:52]: Ah, right... But it seems that those are filtered out already? I'm not able to input anything other than numbers... (except on paste... it will do the first letter of the pasted string, seems to be a bug...)

**Review by zsarnett — DISMISSED** [review]
Damn that looks nice :)

**Comment by balloob** [comment]
This PR is done and I will try not to add more things.

File: src/components/ha-form/compute-initial-ha-form-data.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import { HaFormSchema } from "./types";
+
+export const computeInitialHaFormData = (
+  schema: HaFormSchema[]
+): Record<string, any> => {
+  const data = {};
+  schema.forEach((field) => {
+    if (field.description?.suggested_value) {
+      data[field.name] = field.description.suggested_value;
```
Comment (bramkragten) [inline src/components/ha-form/compute-initial-ha-form-data.ts:9]: Doesn't this mean that we can never not set an optional field with `suggested_value`?

File: src/components/ha-form/ha-form-string.ts [hunk 7]
```diff
@@ -33,48 +37,53 @@ export class HaFormString extends LitElement implements HaFormElement {
...
+    if (changedProps.has("schema")) {
+      this.toggleAttribute("own-margin", !!this.schema.required);
+    }
   }
 
   private _toggleUnmaskedPassword(): void {
     this._unmaskedPassword = !this._unmaskedPassword;
   }
 
   private _valueChanged(ev: Event): void {
```
Comment (bramkragten) [inline src/components/ha-form/ha-form-string.ts:85]: If `value === ""`, should we set it to `undefined`?

File: src/auth/ha-auth-flow.ts [hunk 8]
```diff
@@ -98,16 +146,22 @@ class HaAuthFlow extends litLocalizeLiteMixin(LitElement) {
         `;
       case "error":
         return html`
-          <div class="error">
-            ${this.localize(
+          <ha-alert
+            alert-type="error"
+            .title=${this.localize(
```
Comment (bramkragten) [inline src/auth/ha-auth-flow.ts:151]: I don't think we should use title if we don't have a description, I would put the text in the slot.

**Comment by bramkragten** [comment]
When pressing `<Enter>` while in the password field on the login page, the password is not send

File: src/auth/ha-auth-flow.ts (hunk 8 above)
Comment (balloob) [reply src/auth/ha-auth-flow.ts:151]: Hm interesting. I was reasoning the other way around. Title is emphasized which is what we want error to be.

Comment (bramkragten) [reply src/auth/ha-auth-flow.ts:151]: The element wasn't setup to be used without title, and personally I think it also looks better without the title...

File: src/components/ha-form/ha-form-float.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (balloob) [reply src/components/ha-form/ha-form-float.ts:56]: For integer this code will strip any decimals.

For float, the validty check in mwc-textfield doesn't handle decimals when `type="number"`. Just checks if valid integer. So it will keep saying error.

File: src/dialogs/config-flow/step-flow-form.ts [hunk 9]
```diff
@@ -73,25 +64,13 @@ class StepFlowForm extends LitElement {
             `
           : html`
               <div>
-                <mwc-button
-                  @click=${this._submitStep}
-                  .disabled=${!allRequiredInfoFilledIn}
```
Comment (bramkragten) [inline src/dialogs/config-flow/step-flow-form.ts:78]: Since we now use the `input` event instead of `change` event, we could revert this change? Or do you like the new UX better?

Comment (balloob) [reply src/dialogs/config-flow/step-flow-form.ts:78]: I think I like it more to not disable it.

**Review by bramkragten — APPROVED** [review]
